### ISSUE TITLE
Debugging the connection

### ISSUE THREAD

**Issue by KKSzymanowski** [kdarnok/lwz180#1]
Hi. I was looking for a way to monitor the status of my VRC-W 400 and I came across your repository.
I've tried connecting my ESP32-WROOM-32 with your code to the I2C bus but I don't see any data coming through.

You've mentioned that you used a cheap signal analyzer to see the raw I2C data. Can you point me to the specific hardware you used?

Did you connect to the contacts on top of the unit (under a small cover, next to the pipes) or directly to the contacts for the built-in display?

**Comment by kdarnok** [comment]
Hi, Kuba, I attached everything to the pins at the top of the unit, with the pin assignment described in the manual:

![image](https://github.com/user-attachments/assets/c48bac2a-b523-4aae-b3e8-de07f3478cb9)

The logic analyser I used is one of these Saleae 24MHz 8 channel ones which are sold at various places online. This is the exact one I bought: https://www.roboter-bausatz.de/p/usb-logic-analyzer-24mhz-8ch
With [`sigrok` and `pulseview`](https://sigrok.org/wiki/PulseView) one can capture data and analyse all sorts of protocols.

If you see data coming through using that device, maybe try dumping all i2c frames onto a serial connection on your esp32 so you can test the hardware setup independent of that firmware?

Good luck with your project!

**Comment by KKSzymanowski** [comment]
Yeah, I came across this exact device and it arrived today.
I'm getting the data but the GND line goes high once in a while which is weird. Channels are in the order of pins on the device, meaning:
Ch 0: Clock
Ch 1: Ground
Ch 2: +5V
Ch 3: Data

Have you seen similar things? Do you think it might cause issues? I still should be getting some data because the GND is almost always low.

I'm also seeing some NAKs.

![image](https://github.com/user-attachments/assets/8f47961b-b5c8-455e-9c09-76940d19c9ee)

**Comment by KKSzymanowski** [comment]
> try dumping all i2c frames onto a serial connection on your esp32

Can you point me to some resources explaining how to do that? I have little experience with I2C, especially on the slave side.

**Comment by KKSzymanowski** [comment]
I think I know what might be the issue - when I copied your code I got compilation error on this line:
```c
TWAR = (I2C_SLAVE_ADDR << 1) | 1;  // listen for broadcasts -- all i2c messages are sent using "general calls"
 ```
I just commented it out to make it compile and forgot about it.
Apparently `TWAR` doesn't exist for ESP32 and so I'm unable to listen for broadcasts.

Do you know any alternative mechanism for listening on broadcasts on ESP32 or do I have to buy an actual Arduino board?

**Comment by kdarnok** [comment]
> Yeah, I came across this exact device and it arrived today. I'm getting the data but the GND line goes high once in a while which is weird. Channels are in the order of pins on the device, meaning: Ch 0: Clock Ch 1: Ground Ch 2: +5V Ch 3: Data

i think you may have wired it up incorrectly. this device basically repeatedly measures the voltage (potential difference) on these lines. for that it needs a reference against which to measure. that should be GND -- not the GND of the device, but the GND of the bus. so you should connect GND of the bus to GND of the signal analyser to establish a shared ground.

i would assume that the "fluctuating GND" you're seeing is an artifact of those two GND (bus and signal analyser) being off. +5 doesn't need to be connected iirc.

> Do you know any alternative mechanism for listening on broadcasts on ESP32 or do I have to buy an actual Arduino board?

no, haven't looked much into ESP32 for that purpose. you could try setting the i2c address to `0` -- this might allow you to at least eavesdrop on all the broadcast chatter. but i am not sure it will allow you to send correctly.
you can find a few discussions about this (e.g. https://github.com/espressif/esp-idf/issues/3220), but i would assume you need to select a device that properly support broadcasts (aka general calls).

> Can you point me to some resources explaining how to do that? I have little experience with I2C, especially on the slave side.

i'd suggest you have a look at arduino tutorials and docs -- there's a lot out there. maybe a good starting point could be https://docs.arduino.cc/learn/communication/wire/ and something specifically for the esp32 platform: https://randomnerdtutorials.com/esp32-i2c-communication-arduino-ide/

**Comment by KKSzymanowski** [comment]
> i think you may have wired it up incorrectly.

Yeah, that might've been the case. On my analyzer the last pin is not labeled GND but CND (probably a printing mistake) which misled me into thinking it's  something else. Now I don't get these weird GND spikes.

I did a temporary workaround by connecting the signal analyzer to Raspberry Pi by USB and running sigrok-cli to get the i2c frames, then parsing them in Python and saving the data to the DB to visualize it in Grafana.

I'll probably buy some Arduino or Arduino-compatible hardware to be able to do it properly.

Thanks for your help

**Comment by kubanm3** [comment]
hi @KKSzymanowski did you connect your unit to esp32? i am researching connecting same model as yours to home assistant, probably with esphome.

**Comment by KKSzymanowski** [comment]
@kubanm3 sorry, I missed your comment. I wasn't able to connect directly to ESP32. IIRC the issue was that ESP32 couldn't handle broadcast messages - the TWAR register used [here](https://github.com/kdarnok/lwz180/blob/main/lwz180.ino#L111) doesn't exist on this device, it might be specific to Atmega-based devices.

What I ended up doing is connecting my VRCW-400 to my Arduino Nano v3.0 and sending the gathered data through SPI (because the only I2C bus on Arduino was taken) to my ESP32. It wasn't pretty but it worked 😄

**Comment by roliwa** [comment]
HEllo.     I do not have KNX, but would like to connect directly an Arduino or ESP  to the LWz180 to read all the data.
Do you have a sketch for this ?

**Comment by KKSzymanowski** [comment]
@roliwa You won't be able to connect ESP directly to LWZ because ESP doesn't support reading the broadcast I2C messages as I described in my previous comment.
Your best bet would be connecting it to some Arduino board but make sure you connect through some I2C isolator to avoid the risk of breaking your LWZ.
Take a look at the Wiring section in the readme https://github.com/kdarnok/lwz180#wiring and ignore the "Siemens BCU" part.
This should be a basic I2C setup: connect VCC to VCC, GND to GND and I2C clock and data to appropriate pins on whichever Arduino board you choose.

**Comment by roliwa** [comment]
Thanks you for the answer. I will consider to use an Arduino.I had so fahre Just checked one of the SDP510 DiffPressure Sensors. This worked with an ESP8266. This worked fine. IT IS Standard I2C. So Understand that the Bus IS Not running with Standard clock,...So I follows your Guidance.With this I should bei able to get maybe more Data.Am 13.07.2026 17:13 schrieb Kuba Szymanowski ***@***.***>:KKSzymanowski left a comment (kdarnok/lwz180#1)
@roliwa You won't be able to connect ESP directly to LWZ because ESP doesn't support reading the broadcast I2C messages as I described in my previous comment.
Your best bet would be connecting it to some Arduino board but make sure you connect through some I2C isolator to avoid the risk of breaking your LWZ.
Take a look at the Wiring section in the readme https://github.com/kdarnok/lwz180#wiring and ignore the "Siemens BCU" part.
This should be a basic I2C setup: connect VCC to VCC, GND to GND and I2C clock and data to appropriate pins on whichever Arduino board you choose.

**Comment by KKSzymanowski** [comment]
I haven't tried connecting directly to the sensors in the LWZ/VRC-W unit, they may as well be running normal I2C. The problem with the whole unit is not the clock but the fact that the messages with the data are sent to the address 0 (zero) not some specific address. From what I found ESP32 (not sure about ESP8266) isn't able to listen for messages on address 0.

BTW if you manage to connect, take a look at my PR here: https://github.com/kdarnok/lwz180/pull/2/changes/160157e31243335b2bc5db0f0562973cbd998db0
I added more information about the data being sent by the unit. The data in your unit may not be the same because I have VRC-W 400 and you have LWZ180 but from what I've heard from Stiebel Eltron, these are essentially the same units so it might match.
